crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+ * by appending the (i,j) indices to the seed in the opposite order!
+ *
+ * Where FIPS 203 computs $t = A * s + e$, we use the transpose of "m".
+ */
+static __owur
+int matrix_expand(scalar *out, const uint8_t rho[ML_KEM_RANDOM_BYTES],
+                  int rank, mctx *ctx)
+{
+    uint8_t input[ML_KEM_RANDOM_BYTES + 2];
+    int i, j, ret = 0;
+    EVP_MD_CTX *mdctx = EVP_MD_CTX_new();
+
+    if (mdctx == NULL)
+        goto end;
+    memcpy(input, rho, ML_KEM_RANDOM_BYTES);
+    for (i = 0; i < rank; i++) {
+        for (j = 0; j < rank; j++) {
+            input[ML_KEM_RANDOM_BYTES] = i;
+            input[ML_KEM_RANDOM_BYTES + 1] = j;
+            if (!EVP_DigestInit_ex(mdctx, ctx->shake128_cache, NULL)
+                || !EVP_DigestUpdate(mdctx, input, sizeof(input))
+                || !sample_scalar(out++, mdctx))
+                goto end;
+        }
+    }
+
+    ret = 1;
+end:
+    EVP_MD_CTX_free(mdctx);
+    return ret;
+}
+
+/*
+ * FIPS 203, Section 5.2, Algorithm 14: K-PKE.Encrypt.
+ *
+ * Encrypts a message with given randomness to the ciphertext in |out|. Without
+ * applying the Fujisaki-Okamoto transform this would not result in a CCA
+ * secure scheme, since lattice schemes are vulnerable to decryption failure
+ * oracles.
+ *
+ * Caller passes storage for |y|, |e1| and |u|.
+ */
+static __owur
+int encrypt_cpa(uint8_t *out, const uint8_t *message, const scalar *m,
+                const scalar *t, const uint8_t *randomness, scalar *y,
+                scalar *e1, scalar *u, vinfo_t vinfo, mctx *ctx)
+{
+    uint8_t counter = 0;
+    uint8_t input[ML_KEM_RANDOM_BYTES + 1];
+    scalar e2;
+    scalar v;
+    scalar expanded_message;
+    int rank = vinfo->rank;
+
+    /* FIPS 203 "y" vector */
+    if (!gencbd_vector(y, vinfo->cbd1, &counter, randomness, rank, ctx))
+        return 0;
+    vector_ntt(y, rank);
+    /* FIPS 203 "e1" vector */
+    if (!gencbd_vector(e1, vinfo->cbd2, &counter, randomness, rank, ctx))
+        return 0;
+    memcpy(input, randomness, ML_KEM_RANDOM_BYTES);
+    input[ML_KEM_RANDOM_BYTES] = counter;
+    /* FIPS 203 "e2" scalar */
+    if (!vinfo->cbd2(&e2, input, ctx))
+        return 0;
+    /* FIPS 203 "u" vector */
+    matrix_mult(u, m, y, rank);
+    vector_inverse_ntt(u, rank);
+    vector_add(u, e1, rank);
+    /* FIPS 203 "v" scalar */
+    inner_product(&v, t, y, rank);
+    scalar_inverse_ntt(&v);
+    scalar_add(&v, &e2);
+    /* Extract ciphertext */
+    scalar_decode_1(&expanded_message, message);
+    scalar_decompress(&expanded_message, 1);
+    scalar_add(&v, &expanded_message);
+    vector_compress(u, vinfo->du, rank);
+    vector_encode(out, u, vinfo->du, rank);
+    scalar_compress(&v, vinfo->dv);
+    scalar_encode(out + vinfo->u_vector_bytes, &v, vinfo->dv);
+    return 1;
+}
+
+/*
+ * FIPS 203, Section 5.3, Algorithm 15: K-PKE.Decrypt.
+ */
+static void
+decrypt_cpa(uint8_t *out, const uint8_t *ctext, const scalar *s, scalar *u,
+            vinfo_t vinfo)
```

Comment:
interesting that vinfo_t is a pointer..

Reply:
…which is a pattern we don't use elsewhere.  We typedef the struct not the pointer.

Reply:
In this case (file-local) it is not only implicitly a pointer, it is also a "const pointer". This is quite deliberate *internal* to the file consistency.